pkg/helpers/status.go

```diff
@@ -21,111 +19,97 @@ import (
+	instance.Status.Conditions = []metav1.Condition{}
+	meta.SetResourceCondition(instance, meta.ReadyCondition, metav1.ConditionUnknown, meta.ProgressingReason, "workflow is reconciling...")
 
-	return helper.patchStatus(ctx, instance)
+	return helper.PatchStatus(ctx, instance)
 }
 
-// RollingBack sets the meta.ReadyCondition to 'True' and
-// meta.RollingBack reason and message
-func (helper *StatusHelper) RollingBack(ctx context.Context, instance *v1alpha1.ApplicationGroup) error {
-	meta.SetResourceCondition(instance, meta.ReadyCondition, metav1.ConditionTrue, meta.FailedReason, "rolling back because of failed workflow during upgrade...")
-	meta.SetResourceCondition(instance, meta.DeployCondition, metav1.ConditionTrue, meta.RollingBackReason, "rolling back because of failed workflow during upgrade...")
-
-	return helper.patchStatus(ctx, instance)
+// MarkTerminating sets the meta.ReadyCondition to 'False', with the given
+// meta.Terminating reason and message
+func (helper *StatusHelper) MarkTerminating(instance *v1alpha1.ApplicationGroup) {
+	meta.SetResourceCondition(instance, meta.ReadyCondition, metav1.ConditionFalse, meta.TerminatingReason, "application group is terminating...")
 }
 
-// Progressing resets the conditions of the ApplicationGroup to
-// metav1.Condition of type meta.ReadyCondition with status 'Unknown' and
-// meta.StartingReason reason and message.
-func (helper *StatusHelper) Progressing(ctx context.Context, instance *v1alpha1.ApplicationGroup) error {
-	instance.Status.Conditions = []metav1.Condition{}
-	meta.SetResourceCondition(instance, meta.ReadyCondition, metav1.ConditionUnknown, meta.ProgressingReason, "workflow is reconciling...")
-	meta.SetResourceCondition(instance, meta.DeployCondition, metav1.ConditionUnknown, meta.ProgressingReason, "application group is reconciling...")
+func (helper *StatusHelper) MarkFailed(instance *v1alpha1.ApplicationGroup, err error) {
+	helper.Recorder.Event(instance, "Warning", "ReconcileError", fmt.Sprintf("Failed to reconcile ApplicationGroup %v with Error : %v", instance.Name, err))
+	instance.WorkflowFailed(err.Error())
+}
 
-	return helper.patchStatus(ctx, instance)
+func (helper *StatusHelper) MarkChartPullFailed(instance *v1alpha1.ApplicationGroup, err error) {
+	helper.Recorder.Event(instance, "Warning", "ReconcileError", fmt.Sprintf("Failed to reconcile ApplicationGroup %v with Error : %v", instance.Name, err))
+	instance.ChartPullFailed(err.Error())
 }
 
-// MarkReversing sets the meta.ReadyCondition to 'False', with the given
-// meta.Reversing reason and message
-func (helper *StatusHelper) MarkReversing(ctx context.Context, instance *v1alpha1.ApplicationGroup) error {
-	meta.SetResourceCondition(instance, meta.ReadyCondition, metav1.ConditionFalse, meta.ReversingReason, "application group is reversing...")
-	return helper.patchStatus(ctx, instance)
+func (helper *StatusHelper) MarkTemplateGenerationFailed(instance *v1alpha1.ApplicationGroup, err error) {
```

Comment:
Can we name this `WorkflowTemplate` as to avoid confusion with Helm templating :)

Reply:
Fair enough 😃